File: apps/obsidian-plugin/src/vault/revision.ts
import { createHash } from "node:crypto";

const CONTENT_REVISION_PATTERN = /^sha256:[0-9a-f]{64}$/u;

export function contentRevision(content: string): string {
  return `sha256:${createHash("sha256").update(content, "utf8").digest("hex")}`;
}

export function isContentRevision(value: unknown): value is string {
  return typeof value === "string" && CONTENT_REVISION_PATTERN.test(value);
}

File: crates/grimmore-core/src/revision.rs
//! Content revisions used to prevent stale vault writes.

use sha2::{Digest, Sha256};

/// Returns a portable, algorithm-qualified revision for UTF-8 note content.
#[must_use]
pub fn content_revision(content: &str) -> String {
    let digest = Sha256::digest(content.as_bytes());
    format!("sha256:{}", hex::encode(digest))
}

#[cfg(test)]
mod tests {
    use super::content_revision;

    #[test]
    fn revision_is_deterministic_and_algorithm_qualified() {
        assert_eq!(
            content_revision("Grimmore"),
            "sha256:8b77e81385d6b77d67bb9ea2cb6b269f0401d05f9fa5d9877d75fa28aca9f77b"
        );
    }
}
